File: abstract_ipc.cpp
// Copyright (C) 2020 Adrian Böckenkamp
// This code is licensed under the BSD 3-Clause license (see LICENSE for details).

#ifndef ABSTRACT_IPC_CPP
#define ABSTRACT_IPC_CPP

#include "abstract_ipc.h"

#include <QFile>
#include <QTextStream>
#include <QDebug>

AbstractIPC::AbstractIPC(bool log_debug) : log(log_debug),
  shared_memory(UNIQUE_SHARED_MEMORY_NAME),
  sem_empty(UNIQUE_SEMAPHORE_EMPTY, 1, QSystemSemaphore::AccessMode::Create),
  sem_full(UNIQUE_SEMAPHORE_FULL, 0, QSystemSemaphore::AccessMode::Create)
{
  // Set unique name (key) of shared memory handle, use file if exiting and otherwise hardcoded
  // name:
  file_key = loadKey(SHARED_MEMORY_KEY_FILE);
  if (!file_key.isEmpty()) {
    shared_memory.setKey(file_key);
  }
}

AbstractIPC::~AbstractIPC() { }

QString AbstractIPC::loadKey(const QString path) const
{
  QFile file(path);
  if (file.open(QIODevice::ReadOnly)) {
    QTextStream in_str(&file);
    if (!in_str.atEnd()) {
      QString line = in_str.readLine().trimmed();
      in_str.readLine();
      if (!in_str.atEnd() && log) {
        qDebug() << "Ignoring residual lines in " << path;
      }
      return line;
    }
    file.close();
  }
  return QString();
}

void AbstractIPC::detach()
{
  // See also: https://doc.qt.io/qt-5/qsharedmemory.html#details
  if (!shared_memory.detach() && log) {
    qDebug() << "Unable to detach from shared memory.";
  }
}

#endif // ABSTRACT_IPC_CPP

File: abstract_ipc.py
# -*- coding: utf-8 -*-

# Copyright (C) 2020 Adrian Böckenkamp
# This code is licensed under the BSD 3-Clause license (see LICENSE for details).

import logzero
import PyQt5.QtCore


class AbstractIPC(object):
    """
    Encapsulates code that both the producer and the consumer requires.
    """
    UNIQUE_SHARED_MEMORY_NAME = "MySharedMemoryDefault"
    UNIQUE_SEMAPHORE_EMPTY = "MySemaphoreEmpty"
    UNIQUE_SEMAPHORE_FULL = "MySemaphoreFull"
    SHARED_MEMORY_KEY_FILE = "shared_memory.key"

    def __init__(self, log=True):
        """
        Creates the underlying system ressources.

        :param log: `True` to enable logging using `logzero.logger`, `False` otherwise
        """
        self._log = log
        self._transaction_started = False
        if self._log:
            logzero.logger.info("Using PyQt v" + PyQt5.QtCore.PYQT_VERSION_STR + " and Qt v" +
                                PyQt5.QtCore.QT_VERSION_STR)

        self._file_key = self._load_key(AbstractIPC.SHARED_MEMORY_KEY_FILE)
        self._shared_memory = PyQt5.QtCore.QSharedMemory(self._file_key if self._file_key else
                                                         AbstractIPC.UNIQUE_SHARED_MEMORY_NAME)
        if self._log:
            logzero.logger.debug("Creating shared memory with key=\"" + self._shared_memory.key() + "\" (" +
                                 ("loaded from file)" if self._file_key else "hardcoded)"))
        self._sem_empty = PyQt5.QtCore.QSystemSemaphore(AbstractIPC.UNIQUE_SEMAPHORE_EMPTY, 1,
                                                        PyQt5.QtCore.QSystemSemaphore.Create)
        self._sem_full = PyQt5.QtCore.QSystemSemaphore(AbstractIPC.UNIQUE_SEMAPHORE_FULL, 0,
                                                       PyQt5.QtCore.QSystemSemaphore.Create)

    def _load_key(self, path=SHARED_MEMORY_KEY_FILE):
        """
        Alternatively loads the shared memory's name from a file named `SHARED_MEMORY_KEY_FILE`.

        :param path: Path to file whose first name contains the unique name; the rest is ignored
        :return: Unique name of shared memory or None if such a file did not exist
        """
        # noinspection PyBroadException
        try:
            with open(path) as fp:
                line = fp.readline().strip()
                if fp.readline() and self._log:
                    logzero.logger.warn("Ignoring residual lines in " + path)
                return line
        except:
            pass
        return None
